### ISSUE TITLE
Feature request: size of memory vectors in LSTM

### ISSUE THREAD

**Issue by dkoguciuk** [tensorflow/tensorflow#17510]
## Motivation

There are two different memory vectors used in LSTM cell: internal state vector (_**c**_ symbol in tensorflow convention) and output vector (_**h**_ in tensorflow convention). Let’s consider the possible sizes of both vectors:

- _**c**_ (internal state vector) – has to be the same size for both input and output of a single cell (equation 5 below)

- _**h**_ (output vector) – doesn’t have to be the same (1-4), to be more specific: input size could be arbitrary, but the output size has to be the same as _**c**_ (6)

![lstm_eq](https://user-images.githubusercontent.com/9368849/37099738-671646fa-2221-11e8-9de6-30c46bc36988.png)

The equations are copied from [a great paper](https://arxiv.org/pdf/1506.00019.pdf) of Lipton, Berkowitz and Elkan and again the notation is different: _**s**_=_**c**_).

Why setting different input and output size of _**h**_ vector is it useful? If ‘because we can’ doesn’t satisfy you I can cite [OrderMatters paper](https://arxiv.org/pdf/1511.06391.pdf) (their notation is a bit different _**h**_=_**q**_).

## Current implementation

Currently, in different LSTM cells constructors, there is only one parameter responsible for setting the size of both _**c**_ and _**h**_ vectors and it’s called _**num_units**_. IMO it’s not the best name for this param, but that’s not my point here – there is no way to set different input and output size of _**h**_ vector.

## Proposal

We could leave the _**num_units**_ param for backward compatibility with an integer value and the current behavior. On the other hand, if a user pass a tuple of length 2, we could interpret the param as : (input_size of h, output_size of h). Those two numbers are sufficient for describing both _**c**_ and _**h**_ sizes.

## Implementation

I have already implemented proposed modification for LSTMCell class and it works fine. I could implement the proposed feature for all LSTM-like cells and create PR, but first wanted to ask the community and tensorflowers if it useful form your perspective?

## System information
- **Have I written custom code**: N/A
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: N/A
- **TensorFlow installed from (source or binary)**: N/A
- **TensorFlow version (use command below)**: N/A
- **Python version**: N/A
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**: N/A

**Comment by soloice** [comment]
Hi, friend. 

Are you looking for output projection?
In the follwowing paper, output projection is proposed, which allows `c` and `h` to have different sizes.

>  Hasim Sak, Andrew Senior, and Francoise Beaufays.
>   "Long short-term memory recurrent neural network architectures for
>    large scale acoustic modeling." INTERSPEECH, 2014.

It is already implemented in `tf.nn.rnn_cell.LSTMCell`.

**Comment by dkoguciuk** [comment]
Hello Chong,

I believe the output projection is something different because following the original LTSM equations we do not need any additional matrix to have different input and output sizes of _**h**_. And you're of course right, one can always implement his own RNNCell, but we already have many different version of LSTM Cell. That's why I asked: does TF community needs such LSTM implementation or not? :smiley: 

Best

**Comment by soloice** [comment]
I don't really understand your (input_size of h, output_size of h). 
I glanced over two papers you mentioned above, but didn't find anything related to input size and output size. Did I miss something?

Could you please explain it by pointing out the shapes of all matrices in LSTM update formulae?
Say, if input_size of `h` is `m` and output_size of `h` is `n`, what's the shape of W^{gh} and other matrices?

**Comment by dkoguciuk** [comment]
Hi Chong,

of course! Using the notation from [Lipton, Berkowitz, and Elkan](https://arxiv.org/pdf/1506.00019.pdf):

0. Let's suppose **h^(t-1) is size [m]**
1. W^(gh) will be the size of [n, m] so the g^t will be the size of [n] -- same story for i^t, f^t, and o^t.
2. Both s^(t-1) and s^t will be the size of [n]
3. **h^t will be the size of [n]**

The other params:

4) All b params will be size of [n]
5) All input matrices W^(_x) will be size of [n, input_features_size]
6) x^(t) will be size of [input_features_size]

Does it make sense to you?

**Comment by soloice** [comment]
Probably no.
How come the shapes of h^(t-1) and h^(t) are different! If these two vectors are of different shape, you can't even maintain the RNN loop!

Suppose at a certain time step `t0`, the size of hidden state changed from [m] (the size of h^(t0-1)) to [n] (the size of h^(t0)), how can you run the RNN loop for time step t0+1? The matrix multiplication W^(gh)h^(t0) will cause a mismatch.

Try to write the code and you'll find it out.

**Comment by dkoguciuk** [comment]
Yes, but that's exactly what is proposed in [OrderMatters](https://arxiv.org/pdf/1511.06391.pdf) paper!  After each timestep, the previous output state is concatenated with attention mechanism over the inputs (equation 7).

I've already written all the code and it's working like a charm :smiley:

**Comment by soloice** [comment]
Oh I see it. Thanks for your patience.

In the OrderMatters paper, they say their LSTM has no input, the input state size is `2h`, output state size is `h`, and the recurrent matrices are of shape `2h`-by-`h`.

I think it is just a misnomer.
Actually, you don't need the concatenation stuff: just rewrite the equation (3) as `q_t=LSTM(q_{t-1}, r_t)`, where `q` is the LSTM state of size `h` and `r` is the input vector of size `h`(i.e.: vector `x` in Lipton, Berkowitz, and Elkan's notation), then this LSTM can fit into standard LSTM APIs.

Of course you can post your code here (maybe a github repo link). If your API design is good enough and compatible with the TF library very well, it is worthwhile to include the new RNN cell into TF.

**Comment by dkoguciuk** [comment]
Oh, my! You're right! I can use a standard LSTM API, but why do authors even described this that complicated way? Do you have a clue?

To be sure: I will try to implement the OrderMatters mechanism with standard API and will let you know. And thank you for your suggestion about open-sourcing OrderMatters implementation, I will extract the part of my current project and post it on as a standalone repo.

Getting back to the different input/output sizes of hidden state of LSTM cell: I'm not sure if it's required anymore, but I'll clean up my code and let you know of my implementation.

**Comment by soloice** [comment]
Because the "input" is computed from previous step and cannot be fetched beforehand. The author may want to emphasize that this cell has no additional inputs.

Looking forward to your implementation.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by dkoguciuk** [comment]
I need more time! :smile:

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 75 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by asimshankar** [comment]
Thanks for the discussion @dkoguciuk and @soloice 
Marking as "awaiting response" for now, since IIUC, the requested feature may not be as valuable as originally thought?

**Comment by soloice** [comment]
Agree with you.

**Comment by dkoguciuk** [comment]
Hello, thank you @asimshankar and @soloice! I guess this is sort of an edge case and no one really needs this functionality. I think we could just close the issue.
